torch/csrc/jit/script/init.cpp

```diff
@@ -155,6 +153,40 @@ py::object unpackVariableTensorList(std::vector<at::Tensor> outputs) {
   }
 }
 
+// This is a temporary constructor so that we can write python tests of
+// the executor. It does not have most of the functionality of CompiledFunction
+// such as being able to hold parameters...
+std::shared_ptr<Graph> createGraphByTracing(
+        py::function func,
+        std::vector<tracer::TraceInput> trace_inputs,
+        size_t num_func_inputs) {
+  auto enter_info = tracer::enter(std::move(trace_inputs), 1);
+  py::tuple py_inputs(num_func_inputs);
+  for(size_t i = 0; i < num_func_inputs; ++i) {
+    py_inputs[i] = py::cast(enter_info.second[i]);
+  }
+  auto out = func(*py_inputs);
+  std::vector<autograd::Variable> outputs;
+  if(PyTuple_Check(out.ptr())) {
+    outputs = py::cast<std::vector<autograd::Variable>>(out);
+  } else {
+    outputs.push_back(py::cast<autograd::Variable>(out));
+  }
+  tracer::exit(outputs);
+  auto graph = enter_info.first->graph;
+  EliminateDeadCode(graph);
+  return graph;
+}
+
+static void gatherStateDict(std::vector<at::Tensor*> & values, const Module & m) {
```

Comment:
It includes both buffers and parameters, since `parameters` in C++ includes both. I updated the name to make this clear. Deduplication is handled in python when constructing the trace modules in the first place.

Reply:
Can we document the invariant that says that params are always deduplicated (including occurence in other modules) in C++?